sdk/pinocchio/src/account_info.rs

```diff
@@ -373,17 +382,66 @@ const DATA_SHIFT: u8 = 0;
 
 /// Reference to account data or lamports with checked borrow rules.
 pub struct Ref<'a, T: ?Sized> {
-    value: &'a T,
+    value: NonNull<T>,
     state: NonNull<u8>,
     /// Indicates the type of borrow (lamports or data) by representing the
     /// shift amount.
     borrow_shift: u8,
+    marker: PhantomData<&'a T>,
+}
+
+impl<'a, T: ?Sized> Ref<'a, T> {
+    #[inline]
+    pub fn map<U: ?Sized, F>(orig: Ref<'a, T>, f: F) -> Ref<'a, U>
+    where
+        F: FnOnce(&T) -> &U,
+    {
+        let state = orig.state;
+        let borrow_shift = orig.borrow_shift;
+
+        let value = NonNull::from(f(&*orig));
+
+        // stop the decrement via Drop
+        core::mem::forget(orig);
+
+        Ref {
+            state,
+            value,
+            borrow_shift,
+            marker: PhantomData,
+        }
+    }
```

Comment:
I wonder if we should use a `ManuallyDrop` here – it seems to suite better when transferring memory ownership according to the docs.

```suggestion
    #[inline]
    pub fn map<U: ?Sized, F>(orig: Ref<'a, T>, f: F) -> Ref<'a, U>
    where
        F: FnOnce(&T) -> &U,
    {
        // Avoid decrementing the borrow flag on Drop.
        let orig = ManuallyDrop::new(orig);

        Ref {
            value: NonNull::from(f(&*orig)),
            state: orig.state,
            borrow_shift: orig.borrow_shift,
            marker: PhantomData,
        }
    }
```